[Mobile Payments] Show card's brand and last 4 digits on refunds

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5976
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Adds the Card Brand and Last4 digits to the Refund Confirmation screen for Card Present Payment transactions. This is to help the customer/merchant perform refunds onto the same card which was used for payment.

### Remaining work on 5976
- Test that we fetch the charge when we reach the refund screen, not just display it correctly when passed in #6180
- Do we want to display this for `card` transactions, not just `cardPresent`? I can see arguments both ways but it's straightforward to do. **Leaving this for now, we can add it later if we want to.**
- Stripe gateway support #6181

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
To test, a WooCommerce Payments In Person Payment setup is required. See PdfdoF-D-p2 for info.

1. In the Orders tab, tap an order which has been paid for by a card-present payment
2. Tap `Refund`
3. Select a product to refund or toggle `Refund fees` if required, then tap `Next`
4. Observe that the `Refund Via` row contains card brand and last 4 digits

You can use the [testing card numbers that Stripe provide](https://stripe.com/docs/terminal/references/testing#standard-test-cards) to test different brands/last4 combos with the simulated reader.

N.B. If you follow these instructions and refund immediately after taking payment `Refund Via` screen will not always contain the card details, instead it may just say `WooCommerce Payments`. This is a known issue to be addressed in a future PR.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

Visa | Diners Club
---|---
![Visa card present refund](https://user-images.githubusercontent.com/2472348/153452013-090bbfc1-5724-490f-ba3e-c952369828cc.png) | ![Diners card present refund](https://user-images.githubusercontent.com/2472348/153452018-68bf7777-0a91-4bdc-a2db-e4b06f8dc9b4.png)


https://user-images.githubusercontent.com/2472348/153456499-9a8f0407-a256-4292-854f-32f3ff34a7b5.mp4

@woocommerce/mobile-designers I've gone for two rows here, because card brands can be a little long so combining them with `WooCommerce Payments` on the same line as in the design would lead to them being truncated. Do you think we should reverse the order of the strings here, so that the card details are the title and a little more prominent?

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald committed:
5976 Show card's brand and last 4 digits on refunds

joshheald committed:
5976 Use single localized strings for card details

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/IssueRefundViewModel.swift:
```diff
@@ -94,6 +94,16 @@ final class IssueRefundViewModel {
+    /// Charge related to the order. Used to show card details in the `Refund Via` section.
+    ///
+    private lazy var charge: WCPayCharge? = {
+        guard let resultsController = createWcPayChargeResultsController() else {
```
This is a bit limiting – it's optional because `chargeID` on Order is necessarily optional, and sometimes we don't have the `chargeID` in time when starting a refund. In that case, we're never able to fetch the charge. I'll look to recreate the `resultsController` when the `chargeID` changes, I think.

koke replied:
Maybe we could introduce Combine to `EntityListener`? I imagine if we had that we could observe the order for changes, then `flatMap` that to observe the charge. Completely untested rough example 🙈:

https://gist.github.com/koke/1a6cef8f1a0012719cd757875542a4df

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/RefundConfirmationViewModel.swift:
```diff
@@ -279,3 +287,44 @@ private extension RefundConfirmationViewModel {
+                                     comment: "American Express card brand details. %1$@ is a placeholder for the last 4 digits of the card.")
+        case .diners:
+            return NSLocalizedString("Diners Club •••• %1$@",
+                                     comment: "Diners Club card brand details. %1$@ is a placeholder for the last 4 digits of the card.")
```
Not sure whether these should actually be localizable... or whether we should more strongly discourage changes to them.

[I'm pretty sure that we don't want to reverse the digits in the last4 for RTL languages](https://developer.apple.com/design/human-interface-guidelines/right-to-left/overview/text-and-characters/)

koke replied:
I think it's probably safer to keep them without localization

koke replied:
As I was looking at the Mastercard brand guidelines for the other comment I noticed this:

> The Mastercard name must appear in English only. The Mastercard name must not be translated into other languages nor appear in another alphabet except for specific authorized versions in Chinese (translation), Arabic (transliteration), and Georgian (transliteration).

Probably best to skip translations for these.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Issue Refund/RefundConfirmationViewModelTests.swift:
```diff
@@ -98,11 +101,39 @@ final class RefundConfirmationViewModelTests: XCTestCase {
+        let body = try XCTUnwrap(row.body)
+        XCTAssert(body.contains("MasterCard"))
+        XCTAssert(body.contains("6292"))
+    }
```
Test to add in a future PR: do we fetch the charge? Not doing it here because the fetch mechanism may change somewhat to fix the issues.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/RefundConfirmationViewModel.swift:
```diff
@@ -151,7 +155,11 @@ private extension RefundConfirmationViewModel {
+                return SimpleTextRow(text: details.order.paymentMethodTitle)
+            }
+            return TitleAndBodyRow(title: details.order.paymentMethodTitle,
+                                   body: cardDetails.brand.localizedCardDetails(last4: cardDetails.last4))
```
It would seem slightly neater to have this helper be an extension on the `cardDetails`, so that we didn't need to pass in the `last4`. I might try to improve this, but I want it to also be available for `card` charges if we decide to show this info for them, not just `cardPresent`.

adamzelinski commented:
> Do you think we should reverse the order of the strings here, so that the card details are the title and a little more prominent?

I think it looks good like it is thanks @joshheald

joe-keenan commented:
Hey @joshheald, I think it’d be cleaner if we put the card type in brackets after the number, kinda as you suggested.

My usual method with multiple datapoints on one line is to put the fixed-length one first, and the flexible-length one second. That’s not really an issue here, but it’s a nice consistent methodology to lean on when in doubt.

![image](https://user-images.githubusercontent.com/95003933/153527979-576a864a-bc34-420a-b81a-a9b3971dedb7.png)

joshheald commented:
> My usual method with multiple datapoints on one line is to put the fixed-length one first, and the flexible-length one second. That’s not really an issue here, but it’s a nice consistent methodology to lean on when in doubt.

Thanks Joe, that sounds good and it's really helpful that you shared that insight into your methodology too 👍 ✋

joshheald committed:
5976 Design tweak to card details

joshheald commented:
![Card details with number first](https://user-images.githubusercontent.com/2472348/153568053-8e34938a-10cd-4bfa-9e29-18ae478aa439.png)
Made that change now

koke commented:
I've been testing this and the results are a bit inconsistent. I'm not sure exactly what's happening, but my understanding is that API calls to refresh the order list don't include the charge ID, and end up overwriting the object, even if the single order API call seems to work. Or maybe I'm just not waiting until the order finishes loading content, and then the card details never show up.

koke commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/RefundConfirmationViewModel.swift:
```diff
@@ -279,3 +287,44 @@ private extension RefundConfirmationViewModel {
+            return NSLocalizedString("•••• %1$@ (Visa)",
+                                     comment: "Visa card brand details. %1$@ is a placeholder for the last 4 digits of the card.")
+        case .unknown:
+            return NSLocalizedString("•••• %1$@ (Unknown provider)",
```
Not sure if this is an actual possibility, or just a necessary fallback, but maybe we can just skip the `(Unknown provider)` here? I can imagine that being more confusing than helpful

koke commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/RefundConfirmationViewModel.swift:
```diff
@@ -279,3 +287,44 @@ private extension RefundConfirmationViewModel {
+            return NSLocalizedString("•••• %1$@ (JCB)",
+                                     comment: "JCB card brand details. %1$@ is a placeholder for the last 4 digits of the card.")
+        case .mastercard:
+            return NSLocalizedString("•••• %1$@ (MasterCard)",
```
Looking at [their website](https://brand.mastercard.com/brandcenter/branding-requirements/mastercard.html), it seems the right capitalization is `Mastercard` instead of `MasterCard`.

> When referencing Mastercard® in text, use an uppercase “M” and lowercase “c” with no space between “Master” and “card”. The name must not appear with a capital “C”. Mastercard must appear in the same font as its surrounding text. The name must not be modified in any way. The name may appear in all uppercase letters only if the font style of the user interface or communication also appears in all uppercase letters

joshheald replied:
Thanks for checking and good catch. I've passed this info on to Stripe as it seems like the only one in that object which isn't correct for display.

koke commented:
Another note: I don't think it's a blocker to ship this PR, but we should look into displaying the card brand icons instead of names. Looking at [Mastercard brand guidelines](https://brand.mastercard.com/brandcenter/branding-requirements/mastercard.html) again, they mention "Mastercard" being acceptable next to the last 4 digits only for devices that support text-only:
![Screen Shot 2022-02-14 at 11 11 33](https://user-images.githubusercontent.com/8739/153844339-17650996-76c9-4217-bfa7-527be9213e48.png)

koke reviewed (approved):
Left a few comments, I think it's fine to handle them on future PRs and also improve the loading of charge so the results are more consistent.

joshheald committed:
5976 Changes to card display from code review

We decided not to use localized strings, since card brand names generally have strict requirements around these and display in English everywhere is the safest approach for the time being.

joshheald committed:
5976 Fix inconsistent display of card details

Previously, the refund screen was not displaying the card details consistently, especially missing them for orders where the payment had just been taken.

When loading the order again, the `update(with:)` function was not setting the chargeID we'd recieved from the network on the storage, it was only set when the order was first inserted from a new fetch.

Setting `chargeID` in `update(with:)` fixed these issues, however, @koke's suggestion that they may be caused by the list fetch overwriting the `chargeID` with nil because of the fields on the list fetch not including `meta_data` needed investigation.

I found that they were indeed being overwritten, but then we would fetch fresh data again including the chargeID when entering the Order Details screen before the refund, so in general the chargeID was always present.

To reduce the risk of a race condition or other future bugs here, I changed all Order requests to fetch `meta_data`. With this, I can see that `chargeID` doesn't get incorrectly set to `nil` any more. The `singleOrderExtraFields` was initially added to avoid fetching `line_items` for every order when viewing the list, for performance reasons. We will discuss whether or not this is desirable, and whether we should provide for this in the storage layer with flags to say "don't update these fields".

joshheald committed:
5976 Update Mastercard in test

joshheald committed:
Merge branch 'issue/5976-wcpaycharges-yosemite' into issue/5976-show-last4-and-card-brand-on-refund-confirmation

# Conflicts:
#	Yosemite/Yosemite/Model/Model.swift

joshheald committed:
Merge branch 'issue/5976-show-last4-and-card-brand-on-refund-confirmation' into 5976-improve-consistency-of-fetching-charge

joshheald commented:
> Another note: I don't think it's a blocker to ship this PR, but we should look into displaying the card brand icons instead of names.

I've added #6179

joshheald committed:
Merge pull request #6164 from woocommerce/5976-improve-consistency-of-fetching-charge

[Mobile Payments] Display last4 and card brand on refund screen consistently
